### ISSUE TITLE
The ability to switch between ground mode and floating mode

### ISSUE THREAD

**Issue by KF243** [bicarlsen/easy-biologic#7]
For multi channels' connection, I need to use floating mode instead of ground mode. How can I switch the mode?

**Comment by bicarlsen** [comment]
I added the functionlaity for this in [commit 6a3481](https://github.com/bicarlsen/easy-biologic/commit/6a3481866232e1cd31e15b87d55a2a71a2529d4a). I do not have an SP-300 which is required to test the functionality. If you do, please test it and let me know if it works.

Test Program
```python
import easy_biologic as ebl
import easy_biologic.ec_lib as ecl

bl = ebl.BiologicDevice( <port> )
bl.connect()

bl.harware_configuration()

bl.channel_configuration( <ch> )
bl.set_channel_configuration( <ch>, ecl.ChannelMode.GROUNDED, ecl.ElectrodeConnection.STANDARD )
```

**Comment by KF243** [comment]
easy-biologic 0.3.0.post1 was reinstalled. But TestProgram.py shows the following error.
---
Exception has occurred: ModuleNotFoundError
No module named 'ec_errors'
  File "C:\Users\kfush\data\TestProgram.py", line 2, in <module>
    import easy_biologic.ec_lib as ecl
--
[FG1.zip](https://github.com/bicarlsen/easy-biologic/files/7559394/FG1.zip)
-

**Comment by bicarlsen** [comment]
Ah, sorry, it should be `import easy_biologic.lib.ec_lib as ecl`.

**Comment by bicarlsen** [comment]
Did you download [commit 6a3481](https://github.com/bicarlsen/easy-biologic/commit/6a3481866232e1cd31e15b87d55a2a71a2529d4a) and reinstall the package from the downloaded version?

**Comment by KF243** [comment]
C:\Users\kfush>git clone https://github.com/bicarlsen/easy-biologic.git
Cloning into 'easy-biologic'...
remote: Enumerating objects: 515, done.
remote: Counting objects: 100% (515/515), done.
remote: Compressing objects: 100% (320/320), done.
Receiving objects:  99% (510/515), 10.80remote: Total 515 (delta 308), reused 397 (delta 193), pack-reused 0
MiB/s
Receiving objects: 100% (515/515), 11.39 MiB | 1.19 MiB/s, done.
Resolving deltas: 100% (308/308), done.

I think the package was successfully installed. But TestProgram2.py shows the same error as follows.
Exception has occurred: AttributeError
'BiologicDevice' object has no attribute 'hardware_configuration'
  File "C:\Users\kfush\data\TestProgram2.py", line 11, in <module>
    bl.hardware_configuration

**Comment by bicarlsen** [comment]
After downloading the new commit you need to move into that directory and run `pip install .` to update Python to use that version.  You can run `pip list` afterwards to ensure it was installed correclty. Next to the name it should give the file path to the downloaded version.

**Comment by KF243** [comment]
easy-biologic seems to be in the path correctly as follows.
C:\Users\kfush\AppData\Local\Programs\Python\Python39\Lib\site-packages\easy_biologic>pip list
Package         Version
--------------- -----------
cycler          0.11.0
DateTime        4.3
easy-biologic   0.3.0.post1
fonttools       4.28.1
kiwisolver      1.3.2
matplotlib      3.5.0
numpy           1.21.4
optosigma       1.0.0
packaging       21.2
pandas          1.3.4
Pillow          8.4.0
pip             21.3.1
pyparsing       2.4.7
pyserial        3.5
python-dateutil 2.8.2
pytz            2021.3
scipy           1.7.2
setuptools      58.1.0
setuptools-scm  6.3.2
six             1.16.0
tomli           1.2.2
zope.interface  5.4.0

What is wrong?

**Comment by bicarlsen** [comment]
This means you haven't installed the downloaded version. The new version should be `0.3.1`. Try uninstalling the package, then installing the downloaded version by navigating to its folder and running `pip install .`.

**Comment by KF243** [comment]
I could install the version 0.3.1.
Next TestProgram2.py shows the following message.

Exception has occurred: EcError
ERR_GEN_NOTCONNECTED (-1): No instrument connected.
  File "C:\Users\kfush\data\TestProgram2.py", line 14, in <module>
    bl.channel_configuration( ch )

**Comment by bicarlsen** [comment]
This is because in your program `ch` is defined as a list, but `#channel_configuration` takes in a single channel as an argument. I have fixed some bugs in [commit c6185e](https://github.com/bicarlsen/easy-biologic/commit/c6185e7f82055c1ccc3c6952164e68e40916a956) which you should download and install for further testing.

**Comment by KF243** [comment]
c6185e was committed but it seems no change in the error.

C:\Users\kfush\easy-biologic-c6185e7f82055c1ccc3c6952164e68e40916a956>pip install .
Processing c:\users\kfush\easy-biologic-c6185e7f82055c1ccc3c6952164e68e40916a956
  Preparing metadata (setup.py) ... done
Using legacy 'setup.py install' for easy-biologic, since package 'wheel' is not installed.
Installing collected packages: easy-biologic
  Attempting uninstall: easy-biologic
    Found existing installation: easy-biologic 0.3.1
    Uninstalling easy-biologic-0.3.1:
      Successfully uninstalled easy-biologic-0.3.1
    Running setup.py install for easy-biologic ... done
Successfully installed easy-biologic-0.3.1

---

Exception has occurred: TypeError
list indices must be integers or slices, not list
  File "C:\Users\kfush\data\TestProgram2.py", line 14, in <module>
    bl.channel_configuration( ch )

**Comment by bicarlsen** [comment]
You are likely still passing a list to `channel_configuration`, but only a single channel should be passed. e.g. `bl.channel_configuration( 0 )`.

**Comment by KF243** [comment]
TestProgram21.py shows the following.

Exception has occurred: EcError
ERR_GEN_FUNCTIONFAILED (-6): Function failed.
  File "C:\Users\kfush\data\TestProgram21.py", line 16, in <module>
    bl.set_channel_configuration( 0, ecl.ChannelMode.GROUNDED, ecl.ElectrodeConnection.STANDARD )


[FG3.zip](https://github.com/bicarlsen/easy-biologic/files/7577400/FG3.zip)

**Comment by bicarlsen** [comment]
I added a new upate in [commit 8a016b](https://github.com/bicarlsen/easy-biologic/commit/8a016bc65869a48556c0c115e8f42ba851882e09). See if that helps.

Can you also print out the results from the other calls (i.e. `.hardware_configuration` and `#channel_configuration`) to ensure those are giving sensible results.

**Comment by KF243** [comment]
C:\Users\kfush\easy-biologic-8a016bc65869a48556c0c115e8f42ba851882e09>pip install easy-biologic
Collecting easy-biologic
  Using cached easy_biologic-0.3.0.post1-py3-none-any.whl (23.7 MB)
Installing collected packages: easy-biologic
Successfully installed easy-biologic-0.3.0.post1

This version is OK?
The following .py programs in /Lib/site-packages/easy_biologic folder show problems in VSCode.

In biologic_device.py, "tech" is not defined Pylance(reportUndefinedVariable) [269, 38]
In data_parser.py, Import "ec_lib" could not be resolved Pylance(reportMissingImports) [12, 8]
In ec_lib.py, Import "ec_errors" could not be resolved Pylance(reportMissingImports) [53, 6], "err" is not defined Pylance(reportUndefinedVariable) [639, 12] and [717, 12]

**Comment by bicarlsen** [comment]
No, it should be version 0.3.1.

**Comment by KF243** [comment]
Successfully installed easy-biologic-0.3.1
---
Now, TestProgram21.py can understand bl.hardware_configuration, but shows a new error as follows,

Exception has occurred: EcError
ERR_GEN_NOTCONNECTED (-1): No instrument connected.
  File "C:\Users\kfush\data\TestProgram21.py", line 15, in <module>
    bl.channel_configuration( 0 )

[FG3.zip](https://github.com/bicarlsen/easy-biologic/files/7580332/FG3.zip)

**Comment by bicarlsen** [comment]
Can you try running another command to ensure the device is connected properly. Perhaps print out the device's id with `bl.idn` or something similar. Can you also please print out the results of each call to check if everything is okay.

**Comment by KF243** [comment]
TestProgram22.py shows the followings.
Exception has occurred: EcError
ERR_GEN_FUNCTIONFAILED (-6): Function failed.
  File "C:\Users\kfush\data\TestProgram22.py", line 17, in <module>
    bl.set_channel_configuration( 0, ecl.ChannelMode.GROUNDED, ecl.ElectrodeConnection.STANDARD )
and in terminal,
PS C:\Users\kfush\data>  c:; cd 'c:\Users\kfush\data'; & 'C:\Users\kfush\AppData\Local\Programs\Python\Python39\python.exe' 'c:\Users\kfush\.vscode\extensions\ms-python.python-2021.11.1422169775\pythonFiles\lib\python\debugpy\launcher' '60427' '--' 'c:\Users\kfush\data\TestProgram22.py'
device id = 2

[FG4.zip](https://github.com/bicarlsen/easy-biologic/files/7591875/FG4.zip)

**Comment by bicarlsen** [comment]
Can you please print out the results of `bl.hardware_configuration` and `bl.channel_configuration( <ch> )` as well.

**Comment by KF243** [comment]
I am not sure how to print out the result. But TestPrograms show the attached messages at terminal. 

[FG5.zip](https://github.com/bicarlsen/easy-biologic/files/7594506/FG5.zip)

**Comment by bicarlsen** [comment]
Please try the code in [commit 8c4c30](https://github.com/bicarlsen/easy-biologic/commit/8c4c301a9a894ca70f6afaa2430d4dae57bb3ef3).

**Comment by bicarlsen** [comment]
I was able to test the code in [commit f8e1350](https://github.com/bicarlsen/easy-biologic/commit/f8e135031ee1bf2cca7552a3385b48e69470620c) on our own machine and everything seems to work. Can you please confirm.

**Comment by KF243** [comment]
commit 8c4c30 was uploaded but TestProgram shows "No instrument connected". 
How do you test the code in commit f8e1350 which seems a part of class HardwareConf( c.Structure )?

'idn' in ec_lib.py seems to be the connection id of the device, "2". Why is it 2 instead of others?

**Comment by bicarlsen** [comment]
If it is saying `No instrument connected` you may have a connectivity issue. You can always try running something like `bl.idn` to ensure a connection has been established.

You can use this link to more easily download the code in [commit f8e1350](https://github.com/bicarlsen/easy-biologic/tree/hardware_conf).

The ID of the device isn't important, it makes it so you can communicate with more than one Biologic device from the same computer. If I remember correctly though, the reason it starts at '2' is because the controlling computer is considered device '1'.

**Comment by KF243** [comment]
After restart VSCode, "No instrument connected" was not shown. But TestProgram223.py shows a new error message. 
PS C:\Users\kfush\data> & C:/Users/kfush/AppData/Local/Programs/Python/Python39/python.exe c:/Users/kfush/data/TestProgram223.py
Traceback (most recent call last):
  File "c:\Users\kfush\data\TestProgram223.py", line 12, in <module>
    bl.set_channel_configuration( 0, ecl.ChannelMode.GROUNDED, ecl.ElectrodeConnection.STANDARD )
  File "C:\Users\kfush\AppData\Local\Programs\Python\Python39\lib\site-packages\easy_biologic\device.py", line 280, in set_channel_configuration
    ecl.set_hardware_configuration( self.idn, ch, mode, connection )
  File "C:\Users\kfush\AppData\Local\Programs\Python\Python39\lib\site-packages\easy_biologic\lib\ec_lib.py", line 933, in set_hardware_configuration
    validate( err )
  File "C:\Users\kfush\AppData\Local\Programs\Python\Python39\lib\site-packages\easy_biologic\lib\ec_lib.py", line 1566, in validate
    raise EcError( err )
easy_biologic.lib.ec_errors.EcError: ERR_GEN_FUNCTIONFAILED (-6): Function failed.
PS C:\Users\kfush\data>

Why bl.set_channel_configuration( 0, ecl.ChannelMode.GROUNDED, ecl.ElectrodeConnection.STANDARD ) is not suitable for ecl.set_hardware_configuration( self.idn, ch, mode, connection )?

**Comment by bicarlsen** [comment]
I don't beleive I have **TestProgram223.py**. Are you sure you updated the code to the new version?

**Comment by KF243** [comment]
Sorry. Find TestProgram223.py in the attachment.

[FG6.zip](https://github.com/bicarlsen/easy-biologic/files/7614715/FG6.zip)

**Comment by bicarlsen** [comment]
It appears you are calling the functions correctly. I sgugest printing out each call to make sure you are getting sensible results. Please also make sure you have [commit f8e1350](https://github.com/bicarlsen/easy-biologic/tree/hardware_conf) installed.

**Comment by KF243** [comment]
How can I print out each call using VSCode?

**Comment by bicarlsen** [comment]
Just by wrapping the result in a `print()` statment. e.g. `print( bl.idn )`

**Comment by KF243** [comment]
Terminal comment of TestProgram224.py was printed out.

[FG7.zip](https://github.com/bicarlsen/easy-biologic/files/7617301/FG7.zip)

**Comment by bicarlsen** [comment]
The output doesn't seem to have any output, but is just the same program.

**Comment by KF243** [comment]
Sorry. Check the .txt file in re-attached.

[FG7.zip](https://github.com/bicarlsen/easy-biologic/files/7622197/FG7.zip)

**Comment by bicarlsen** [comment]
It appears you don't have the correct version of the code installed. Please try removing the package, and reinstalling from [commit f8e1350](https://github.com/bicarlsen/easy-biologic/tree/hardware_conf). 

You can ensure the correct version of the code is installed by printing out `bl.channel_configuration( 0 )` and ensureing it looks like `HardwareConfiguration(connection=<ElectrodeConnection.STANDARD: 0>, mode=<ChannelMode.FLOATING: 1>` instead of `<easy_biologic.lib.ec_lib.HardwareConf object at 0x0000028972C99040>`

**Comment by KF243** [comment]
This time, zip file was used for installing commit f831350 instead of "git clone https://github.com/bicarlsen/easy-biologic.git". Anyways, the followings are shown and it seems to be OK.
PS C:\Users\kfush\data> & C:/Users/kfush/AppData/Local/Programs/Python/Python39/python.exe c:/Users/kfush/data/TestProgram224.py
6- None
7- {0: HardwareConfiguration(connection=<ElectrodeConnection.STANDARD: 0>, mode=<ChannelMode.GROUNDED: 0>), 1: HardwareConfiguration(connection=<ElectrodeConnection.STANDARD: 0>, mode=<ChannelMode.GROUNDED: 0>)}
11- HardwareConfiguration(connection=<ElectrodeConnection.STANDARD: 0>, mode=<ChannelMode.GROUNDED: 0>)
12- None
13- HardwareConfiguration(connection=<ElectrodeConnection.STANDARD: 0>, mode=<ChannelMode.GROUNDED: 0>)
14- None
PS C:\Users\kfush\data>

**Comment by KF243** [comment]
I want to use in FLOATNG mode instead of GROUND mode.
However, currently, channel mode cannot be changed to FLOATING.

[FG8.zip](https://github.com/bicarlsen/easy-biologic/files/7623341/FG8.zip)

**Comment by bicarlsen** [comment]
Great! It appears to be almost working now.

I think this may have been a bug where I accientally switched the tqo parameters somewhere. Please try [commit 541b183](https://github.com/bicarlsen/easy-biologic/tree/hardware_conf).

**Comment by KF243** [comment]
It becomes worse. Check the result text.
Anyway, which do I download a new commit, via zip file or via git? 
The errors are not same.

[TestProgram226_printout_version2.txt](https://github.com/bicarlsen/easy-biologic/files/7624661/TestProgram226_printout_version2.txt)

**Comment by bicarlsen** [comment]
It appears that the output you sent and the previous one are the same, so I think the code didn't update.

The best way to update the code is to use `git pull <remote_name> <branch_name>` then ensure you have `checkout`ed the correct branch. In this case you want to use the brnach `hardware_conf` so the command will look comsething like `git pull origin hardware_conf & git checkout hardware_conf`. However I leave it to you to ensure you are one the correct brnach by comparing the commit ID. You can do this using `git log`. If these are unfamiliar to you, I suggest doing a tutorial on basic git commands and usage.

**Comment by KF243** [comment]
TestProgram226.py with easy-biologic downloaded (correctly, installed via .zip file) the latest hardware_conf (b8bfd1c) results in showing EM_b8bfd1c_TestProgram226py.txt.
Then I tried to reinstall commit f8e1350 from this issue, but it seems to be changed and the same as b8bfd1c.

Easy-biologic-hardware_conf_OK, which is the previous commit downloaded via .zip  previous commit, shows the result of OK_TestProgram226py.txt as previously reported, although I cannot find and download the same commit from your github now.

[FG9.zip](https://github.com/bicarlsen/easy-biologic/files/7646112/FG9.zip)

[technologies.zip](https://github.com/bicarlsen/easy-biologic/files/7646175/technologies.zip)

[OK_without_technologiesInEasy-Biologic.zip](https://github.com/bicarlsen/easy-biologic/files/7646186/OK_without_technologiesInEasy-Biologic.zip)

**Comment by bicarlsen** [comment]
Sorry about that. I fixed the errors in [commit 14615aa](https://github.com/bicarlsen/easy-biologic/tree/14615aac24343813148a5ce7e5637f8c4720b305).

**Comment by KF243** [comment]
FLOATING mode are available now. Thank you.
I will use this "set_channel_configuration" for SECM operation.

**Comment by bicarlsen** [comment]
Great :) Glad we finally got it working.

I've included the changes in [version 0.3.1 of the package](https://pypi.org/project/easy-biologic/)
